File: 240/search-a-2d-matrix-ii.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
    if (!matrix.length)
        return false;

    let [m, n] = [matrix.length, matrix[0].length];

    function search(x, y) {
        if (x >= 0 && x < m && y >= 0 && y < n) {
            if (matrix[x][y] == target)
                return true;
            else if (matrix[x][y] > target)
                return search(x-1, y);
            else
                return search(x, y+1);
        } else return false;
    }
    return search(m-1, 0);
    
};

File: 074/SearchA2DMatrixII.kt
class SearchA2DMatrixII {

    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        return searchMatrixUtil(matrix, target, matrix.size - 1, 0)
    }

    private fun searchMatrixUtil(matrix: Array<IntArray>, target: Int, x: Int, y: Int): Boolean {
        if (matrix.isEmpty())
            return false

        val m = matrix.size
        val n = matrix[0].size

        return if (x in 0 until m && y in 0 until n) {
            if (matrix[x][y] == target)
                true
            else if (matrix[x][y] > target)
                searchMatrixUtil(matrix, target, x - 1, y)
            else
                searchMatrixUtil(matrix, target, x, y + 1)
        } else
            false
    }
}

fun main(args: Array<String>) {
    val matrix = arrayOf(intArrayOf(-5))
    val target = -5
    println(SearchA2DMatrixII().searchMatrix(matrix, target))
}
